src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,181 @@
+      // Assert that It obeys the STL forward iterator concept
+      ASSERT_INPUT_ITERATOR(IteratorT);
+      for(IteratorT it = initialData_begin; it != initialData_end; ++it) {
+        append(*it);
+      }
+    }
+
+    const UInt maxCapacity;
+    const std::string ID; //name of this object
+    const int DEBUG;
+
+    size_t size() const {
+      NTA_ASSERT(buffer_.size() <= maxCapacity);
+      return buffer_.size();
+    }
+
+
+    /** append new value to the end of the buffer and handle the 
+       "overflows"-may pop the oldest element if full. 
+      */
+    void append(T newValue) {
+      if(size() < maxCapacity) {
+        buffer_.push_back(newValue);
+      } else {
+        buffer_[idxNext_] = newValue;
+      }
+      idxNext_ = (idxNext_ +1 ) %maxCapacity;
+    }
+
+
+      /** like append, but return the dropped value if it was dropped.
+        :param T newValue - new value to append to the sliding window
+        :param T* - a return pass-by-value with the removed element,
+          if this function returns false, this value will remain unchanged.
+        :return bool if some value has been dropped (and updated as 
+          droppedValue) 
+      */
+      bool append(T newValue, T* droppedValue) {
+        //only in this case we drop oldest; this happens always after
+        //first maxCap steps ; must be checked before append()
+        const bool isFull = (buffer_.size()==maxCapacity);
+        if(isFull) {
+          *droppedValue = buffer_[idxNext_];
+        }
+        append(newValue);
+        return isFull;
+      }
+
+
+      /**
+        :return unordered content (data ) of this sl. window; 
+          call getLinearizedData() if you need them oredered from 
+          oldest->newest
+        This direct access method is fast.
+      */
+      const std::vector<T>& getData() const {
+        return buffer_;
+      }
+
+
+      /** linearize method for the internal buffer; this is slower than 
+        the pure getData() but ensures that the data are ordered (oldest at
+        the beginning, newest at the end of the vector
+        This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+        :return new linearized vector
+      */
+      std::vector<T> getLinearizedData() const {
+        std::vector<T> lin;
+        lin.reserve(buffer_.size());
+
+        //insert the "older" part at the beginning
+        lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_));
+        //append the "newer" part to the end of the constructed vect
+        lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_);
+        return lin;
+      }
+
+
+      bool operator==(const SlidingWindow& r2) const {
+        const bool sameSizes = (this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity);
+        return sameSizes && std::equal(this->buffer_.cbegin(), this->buffer_.cend(), r2.getData().cbegin()); //also content must be same
```

Comment:
It's not correct to compare the vectors. Like you mentioned, just using `==` with two vectors already does that. Two `SlidingWindow`s can be logically different while having the same vector (if the `idxNext_` is different). Similarly, two `SlidingWindow`s can be logically the same while having different vectors. You need to compare in logical order, not in the order that they're stored in the vector, and the easiest way to do that is with the MovingWindow's `[]` operator.